superset/db_engine_specs.py

```diff
@@ -1185,6 +1193,375 @@ def get_all_datasource_names(cls, db, datasource_type: str) \
                 schema=row['table_schema'], table=row['table_name']))
         return datasource_names
 
+    @classmethod
+    def _build_column_hierarchy(cls,
+                                columns: List[dict],
+                                parent_column_types: List[str],
+                                column_hierarchy: dict) -> None:
+        """
+        Build a graph where the root node represents a column whose data type is in
+        parent_column_types. A node's children represent that column's nested fields
+        :param columns: list of columns
+        :param parent_column_types: list of data types that decide what columns can
+               be root nodes
+        :param column_hierarchy: dictionary representing the graph
+        """
+        if len(columns) == 0:
+            return
+        root = columns.pop(0)
+        root_info = {'type': root['type'], 'children': []}
+        column_hierarchy[root['name']] = root_info
+        while columns:
+            column = columns[0]
+            # If the column name does not start with the root's name,
+            # then this column is not a nested field
+            if not column['name'].startswith('{}.'.format(root['name'])):
+                break
+            # If the column's data type is one of the parent types,
+            # then this column may have nested fields
+            if str(column['type']) in parent_column_types:
+                cls._build_column_hierarchy(columns, parent_column_types,
+                                            column_hierarchy)
+                root_info['children'].append(column['name'])
+                continue
```

Comment:
I'm confused by this `continue`. If you `continue` here won't you end up in an infinite loop? Ah, it recurses and modified `columns` in place, so on the next step of the loop `columns` has changed, is that right?

Reply:
Yes yes.